distribution/tools/keystore-cli/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java

```diff
@@ -59,42 +50,35 @@ protected void executeCommand(Terminal terminal, OptionSet options, Environment
 
         final KeyStoreWrapper keyStore = getKeyStore();
 
-        final Closeable closeable;
         final CheckedFunction<String, char[], IOException> valueSupplier;
         if (options.has(stdinOption)) {
-            final BufferedReader stdinReader = new BufferedReader(new InputStreamReader(getStdin(), StandardCharsets.UTF_8));
             valueSupplier = s -> {
-                try (CharArrayWriter writer = new CharArrayWriter()) {
-                    int c;
-                    while ((c = stdinReader.read()) != -1) {
-                        if ((char) c == '\r' || (char) c == '\n') {
-                            break;
-                        }
-                        writer.write((char) c);
+                try {
+                    return terminal.readSecret("");
+                } catch (Exception e) {
+                    if (e == EMPTY_LINE_EXCEPTION) {
+                        return new char[0];
```

Comment:
This does not actually occur on an empty line, for example when there are two successive newlines in the input stream: `val1\n\nval3\n`. We get this exception when the stream is closed. So if a user runs `echo bar1 | elasticsearch-keystore add-string -x foo1 foo2 foo3`, we'll  blank out the foo2 and foo3 settings instead of throwing an error. I'm not sure this is what we want to do in this case. If the user wants to zero out those values, they should be able to run something like `(echo bar1 ; echo "" ; echo "") | elasticsearch-keystore add-string -x foo1 foo2 foo3`.